File: lib/data/autologin.dart
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

import '../services/log.dart';

class AutoLogin {
  AutoLogin._();

  static final AutoLogin instance = AutoLogin._();

  static const _scope = 'AutoLogin';
  static const _key = 'vault.master';

  // first_unlock, not unlocked: auto-unlock must work when the app is launched
  // into the background after a reboot.
  static const _storage = FlutterSecureStorage(
    iOptions: IOSOptions(accessibility: KeychainAccessibility.first_unlock),
    mOptions: MacOsOptions(accessibility: KeychainAccessibility.first_unlock),
  );

  bool? _available;
  String? _unavailableReason;

  String? get unavailableReason => _unavailableReason;

  /// Probes with a write+delete, not a read: on a sandboxed macOS build reading
  /// a missing key succeeds while writing fails with -34018.
  Future<bool> isAvailable() async {
    final cached = _available;
    if (cached != null) return cached;
    const probeKey = 'keystore.probe';
    try {
      await _storage.write(key: probeKey, value: 'ok');
      await _storage.delete(key: probeKey);
      _unavailableReason = null;
      Log.info(_scope, 'keystore available (write probe passed)');
      return _available = true;
    } catch (error, stackTrace) {
      _unavailableReason = describeFailure(error);
      Log.error(_scope, 'keystore probe failed: $error', stackTrace);
      Log.warn(_scope, 'auto-unlock disabled: $_unavailableReason');
      return _available = false;
    }
  }

  Future<bool> isEnabled() async {
    try {
      return await _storage.containsKey(key: _key);
    } catch (error, stackTrace) {
      Log.error(_scope, 'containsKey failed', stackTrace);
      Log.error(_scope, error);
      return false;
    }
  }

  Future<String?> readPassword() async {
    try {
      return await _storage.read(key: _key);
    } catch (error, stackTrace) {
      Log.error(_scope, 'read failed: $error', stackTrace);
      return null;
    }
  }

  Future<void> enable(String password) async {
    try {
      await _storage.write(key: _key, value: password);
      Log.info(_scope, 'auto-unlock enabled');
    } catch (error, stackTrace) {
      Log.error(_scope, 'write failed: $error', stackTrace);
      throw AutoLoginException(describeFailure(error));
    }
  }

  /// Never throws: revoking must always succeed from the user's side.
  Future<void> disable() async {
    try {
      await _storage.delete(key: _key);
      Log.info(_scope, 'auto-unlock disabled');
    } catch (error, stackTrace) {
      Log.error(_scope, 'delete failed: $error', stackTrace);
    }
  }

  static String describeFailure(Object error) {
    final text = '$error';
    if (text.contains('-34018')) {
      return 'macOS keychain refused access (-34018). This build is signed '
          'ad-hoc; pick a signing Team and add Keychain Sharing in Xcode — '
          'see the comment in macos/Runner/DebugProfile.entitlements.';
    }
    if (text.contains('-25291')) {
      return 'No keychain is available (-25291).';
    }
    if (text.contains('-25300')) {
      return 'Keychain item not found (-25300).';
    }
    if (text.toLowerCase().contains('libsecret') ||
        text.toLowerCase().contains('secret service')) {
      return 'No secret service on this Linux session — install libsecret and '
          'run a keyring daemon.';
    }
    return text;
  }
}

class AutoLoginException implements Exception {
  const AutoLoginException(this.message);
  final String message;
  @override
  String toString() => message;
}

File: shared/src/commonMain/kotlin/com/example/ohmyssh/data/AutoLogin.kt
package com.example.ohmyssh.data

import com.example.ohmyssh.platform.SecureStorage
import com.example.ohmyssh.services.Log

object AutoLogin {
    private const val SCOPE = "AutoLogin"
    private const val KEY = "vault.master"

    private var available: Boolean? = null

    var unavailableReason: String? = null
        private set

    var unavailableIsExpected: Boolean = false
        private set

    /// Probes with a write+delete, not a read: on a sandboxed macOS build
    /// reading a missing key succeeds while writing fails with -34018.
    suspend fun isAvailable(): Boolean {
        available?.let { return it }
        val probeKey = "keystore.probe"
        return try {
            SecureStorage.write(probeKey, "ok")
            SecureStorage.delete(probeKey)
            unavailableReason = null
            unavailableIsExpected = false
            Log.info(SCOPE, "keystore available (write probe passed)")
            available = true
            true
        } catch (error: Exception) {
            unavailableIsExpected = isExpectedFailure(error)
            unavailableReason = describeFailure(error)
            if (!unavailableIsExpected) {
                Log.error(SCOPE, "keystore probe failed: $error", error)
            }
            Log.warn(SCOPE, "auto-unlock off: $unavailableReason")
            developerHint(error)?.let { Log.info(SCOPE, it) }
            available = false
            false
        }
    }

    suspend fun isEnabled(): Boolean {
        if (available == false) return false
        return try {
            SecureStorage.containsKey(KEY)
        } catch (error: Exception) {
            report("containsKey failed", error)
            false
        }
    }

    suspend fun readPassword(): String? {
        if (available == false) return null
        return try {
            SecureStorage.read(KEY)
        } catch (error: Exception) {
            report("read failed", error)
            null
        }
    }

    suspend fun enable(password: String) {
        try {
            SecureStorage.write(KEY, password)
            Log.info(SCOPE, "auto-unlock enabled")
        } catch (error: Exception) {
            report("write failed", error)
            throw AutoLoginException(describeFailure(error))
        }
    }

    suspend fun disable() {
        if (available == false) return
        try {
            SecureStorage.delete(KEY)
            Log.info(SCOPE, "auto-unlock disabled")
        } catch (error: Exception) {
            report("delete failed", error)
        }
    }

    private fun report(what: String, error: Exception) {
        if (isExpectedFailure(error)) {
            Log.warn(SCOPE, "$what: ${describeFailure(error)}")
            return
        }
        Log.error(SCOPE, "$what: $error", error)
    }

    fun isExpectedFailure(error: Any): Boolean {
        val text = "$error".lowercase()
        return text.contains("-34018") ||
            text.contains("-25291") ||
            text.contains("-25308") ||
            text.contains("libsecret") ||
            text.contains("secret service") ||
            text.contains("secret-tool") ||
            text.contains("no keystore backend")
    }

    fun describeFailure(error: Any): String {
        val text = "$error"
        if (text.contains("-34018")) {
            return "macOS denies keychain access to this build (-34018) because it " +
                "is signed ad-hoc."
        }
        if (text.contains("-25291")) return "No keychain is available (-25291)."
        if (text.contains("-25308")) return "The keychain is locked (-25308)."
        if (text.contains("-25300")) return "Keychain item not found (-25300)."
        val lower = text.lowercase()
        if (lower.contains("libsecret") || lower.contains("secret service") ||
            lower.contains("secret-tool")
        ) {
            return "No secret service on this Linux session — install libsecret and " +
                "run a keyring daemon."
        }
        if (lower.contains("no keystore backend")) {
            return "This platform has no keystore backend."
        }
        return text
    }

    private fun developerHint(error: Any): String? {
        if (!"$error".contains("-34018")) return null
        return "to enable auto-unlock locally, sign the app with a Team and the " +
            "keychain-access-groups entitlement."
    }
}

class AutoLoginException(override val message: String) : Exception(message) {
    override fun toString(): String = message
}
